vars/yamlSubstituteVariables.groovy

```diff
@@ -0,0 +1,314 @@
+     */
+    'outputManifestFile',
+    /**
+     * The {@code String} path of the Yaml file containing the variables' values to use as a replace in the manifest file. <br>
+     * Defaults to "manifest-variables.yml" if not specified otherwise
+     */
+    'variablesFile'
+]
+
+@Field Set PARAMETER_KEYS = STEP_CONFIG_KEYS
+
+/**
+ * Step that substitutes variables in a given YAML file with those specified in a another. The format to reference a variable
+ * in the YAML file is to use double parentheses {@code ((} and {@code ))}, e.g. {@code ((variableName))}.<br>
+ * A declaration of a variable and assignment of its value is simply done as a property in the variables YAML file.
+ * <p>
+ * The format follows <a href="https://docs.cloudfoundry.org/devguide/deploy-apps/manifest-attributes.html#variable-substitution">Cloud Foundry standards</a>.
+ * <p>
+ * The step is activated by the presence of both a {@code manifest.yml} and a variables file. Names of both files are configurable.
+ * <p>
+ * Usage: yamlSubstituteVariables manifestFile: "manifest.yml", variablesFile: "manifest-variables.yml"
+ *
+ * @param arguments - the map of arguments.
+ */
+@GenerateDocumentation
+void call(Map<String, String> arguments) {
+    handlePipelineStepErrors (stepName: STEP_NAME, stepParameters: arguments) {
+        def script = checkScript(this, arguments)  ?: this
+
+        // load default & individual configuration
+        Map config = ConfigurationHelper.newInstance(this)
+                                        .loadStepDefaults()
+                                        .mixinStepConfig(script.commonPipelineEnvironment, STEP_CONFIG_KEYS)
+                                        .mixinStageConfig(script.commonPipelineEnvironment, arguments.stageName ?: env.STAGE_NAME, STEP_CONFIG_KEYS)
+                                        .mixin(arguments, PARAMETER_KEYS)
+                                        .use()
+
+        String manifestFilePath = config.manifestFile ?: "manifest.yml"
+        String variablesFilePath = config.variablesFile ?: "manifest-variables.yml"
+        String outputFilePath = config.outputManifestFile ?: manifestFilePath
+
+        // Parameter for tests only. No public API.
+        // This is to skip manifest file deletion in tests, which
+        // would be a nuisance, as it alters the sources repository.
+        Boolean skipDeletion = arguments.skipDeletion ?: false
+
+        File manifestFile = new File(manifestFilePath)
+        File variablesFile = new File(variablesFilePath)
+
+        if (!manifestFile.exists()) {
+            echo "[YamlSubstituteVariables] Could not find YAML file at ${manifestFilePath}. Skipping variable substitution."
+            return
+        }
+
+        if (!variablesFile.exists()) {
+            echo "[YamlSubstituteVariables] Could not find variable substitution file at ${variablesFilePath}. Skipping variable substitution."
```

Comment:
Not sure if simply returning in case the file does not exist is a good pattern. With such an approach we silently skip the attempt to replace content. Wouldn't it be better to fail with an exception? I mean, the variable substitution step is introduced by a pipeline author for a reason. In case a certain file does not exist we should fail here rather than failing later.

Reply:
See above.